src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AbstractFrameCheck.java

```diff
@@ -0,0 +1,738 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
+
+/**
+ * Abstract class for checks which require to work with and identify many frames.
+ */
+@FileStatefulCheck
+public abstract class AbstractFrameCheck extends AbstractCheck {
```

Comment:
I question moving this to an abstract class instead of an utility. If people want to use this functionality, they are forced to extend the class and can't extend any custom class of their own. This then can't be used for javadoc checks.

Reply:
I get your point, @romani thinks similar too ( https://github.com/checkstyle/checkstyle/pull/8013#discussion_r405513737).
Thanks to your example, I better understand the benefits of such design, 

I will propose a new design making use of a `FrameHelper` class instead.

Reply:
Done.